streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStore.java

```diff
@@ -29,21 +29,33 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.NavigableMap;
-import java.util.Set;
+import java.util.NavigableSet;
 import java.util.TreeMap;
 import java.util.TreeSet;
+import java.util.concurrent.ConcurrentSkipListMap;
 
 public class InMemoryKeyValueStore implements KeyValueStore<Bytes, byte[]> {
 
     private static final Logger LOG = LoggerFactory.getLogger(InMemoryKeyValueStore.class);
 
     private final String name;
-    private final NavigableMap<Bytes, byte[]> map = new TreeMap<>();
+    private final boolean copyOnRange;
+    private final NavigableMap<Bytes, byte[]> map;
+
     private volatile boolean open = false;
     private long size = 0L; // SkipListMap#size is O(N) so we just do our best to track it
 
+    // for tests-only
     public InMemoryKeyValueStore(final String name) {
+        this(name, true);
+    }
+
+    public InMemoryKeyValueStore(final String name, final boolean copyOnRange) {
         this.name = name;
+        this.copyOnRange = copyOnRange;
+        // if we do not need to copy on range, then we should use concurrent skiplist map
+        // to avoid concurrent modificiation exception
+        this.map = copyOnRange ? new TreeMap<>() : new ConcurrentSkipListMap<>();
```

Comment:
@guozhangwang , I saw you added the `ConcurrentSkipListMap` back for not copy-on-range case, but at the same time, we still have `synchronized` lock for map operation methods. That seems over-locked, and should have worse performance. What do you think?